chore: add testing development notes (#1343)
verdaccio/verdaccio: test/README.md

## How to test Verdaccio

> Welcome to the testing folder at Verdaccio. This document aims to help you to go through how Verdaccio can be tested.

Reviewer: ```suggestion
Welcome to the testing folder at Verdaccio. This document aims to help you understand how Verdaccio should be tested.
```

> First of all we must explain the stack, we use `jest` and other tools such as `supertest` to be able to test the API and `puppeteer` for End-to-End testing.

Reviewer: ```suggestion
First of all, we should explain the testing frameworks being used. We use `jest` and other tools such as `supertest` to be able to test the API, and `puppeteer` for End-to-End testing.
```

We go along with the following rules in order to be consistent with all tests which will make your code review smooth and fast:

* We **type** all our tests. eg `const foo: number = 3;`

> * **Each test should be as small as possible**: You should try the `test()` block to test only one thing and do not depend on other tests. If the test requires different steps, group them with a `describe()` block.

Reviewer: ```suggestion
* **Each test should be as small as possible**: You should use the `test()` block to test only one thing and do not depend on other tests. If the test requires different steps, group them with a `describe()` block.
```

> * All `test()` **headers titles** should begin with `test('should test ...)`: For consistency with reporting tools, this makes it easier to match the test with the feature needed to be tested.

Reviewer: ```suggestion
* All `test()` **headers titles** should begin with `test('should test ...')`: For consistency with reporting tools, this makes it easier to match the test with the feature needed to be tested.
```

* **Any mock data** should be located in the `partials` folder in each section.
* Use `yaml` for **configuration files examples** instead of JSON.
* If you use a **file based mock storage**, it should be located in the `store` folder in each section.
* All tests **MUST NOT** rely on external sources and must be able to run them **offline**.
* Tests **must run on the following Operating Systems**: Unix (Mac, Linux) and Windows (7 -> latest).
* If you are creating mock data file which use the state and need a clean state, use `rimraf` to remove folders.

## Testing sections

> Verdaccio testing is being separated in 3 sections, each of them has different setup and scope, the most important is the **unit test**. All sections have custom **jest configuration files**.

Reviewer: ```suggestion
Verdaccio testing is split in 3 sections, each of them has different setup and scope. The most important is the **unit test**. All sections have custom **jest configuration files**.
```

If you are adding new tests, comply with the following:

> * If you add a new API endpoint, the  unit test and functional test are mandatory.

Reviewer: ```suggestion
* If you add a new API endpoint, unit and functional tests are mandatory.
```

* If you add a utility, unit test is mandatory.
* If you are adding a new web API endpoint, the unit test, functional test and if such endpoint has new changes in the UI, E2E test is also mandatory.
* If you add or refactor a core class, unit test is mandatory.

> * If you fix a bug, you **must** add a new `test()` block to prove that the patch fixes te bug.

Reviewer: ```suggestion
* If you fix a bug, you **must** add a new `test()` block to prove that the patch fixes the bug.
```

### Unit test

> Unit test aims to test the CLI API and the Web API. The configuration file is located at `jest.config.js`.

Reviewer: ```suggestion
Unit tests aim to test the CLI API and the Web API. The configuration file is located at `jest.config.js`.
```

All tests must be included in the `test/functional/index.spec.js` file, which bootstraps the whole process. There is only one spec file and **must be only one**.

The jest configuration file is defined in `test/jest.config.functional.js`. The configuration will create a custom environment launching 3 Verdaccio servers with different configurations.

The servers are linked as follows: 

* Server 1 
  * -> Server 2  
  * -> Server 3
* Server 2 
  * -> Server 1
* Server 3 
  * -> Server 2
  * -> Server 1
* Express app: (if you need to emulate any external endpoint, use the express app)  

Server 1 runs on port `55551`, Server 2 on port `55552` and Server 3 on port `55553`.

> If you have the need to increase the number of servers running, it is possible, but please discuss with the team before you go in that path.

#### Adding a new block

To add a new feature you need to export the feature as a function that take as an argument any of the servers you want to interact. 

```js
// newFeature.js

export default function(server) {

  describe('package access control', () => {
    test('should ...', (done) => {
      done();
    });
  });
  
}
```

> Then, import the feature and run the function within the main `describe` block.  

Reviewer: ```suggestion
Then import the feature and run the function within the main `describe` block.  
```

```js
// index.spec.js

import newFeature from './newFeature';

describe('functional test verdaccio', function() {
  // test are fast, but do not change this time out, 10 seconds should be more than enough
  jest.setTimeout(10000);

>   // servers are accessed vi a global jest state.

Reviewer: ```suggestion
  // servers are accessed via a global jest state.
```

  const server1: IServerBridge = global.__SERVERS__[0];
  const server2: IServerBridge = global.__SERVERS__[1];
  const server3: IServerBridge = global.__SERVERS__[2];
  const app = global.__WEB_SERVER__.app;

  // include as much servers you need
  newFeature(server1, server2, server3);
});
```

> Functional test run over ones single file, thus, it is not possible at this stage to run test individually.

Reviewer: ```suggestion
Functional tests run over one single file, thus, it is not possible at this stage to run tests individually.
```

Reviewer: it's possible however to use jest's `test.only` or `test.skip` and using that to selectively run or ignore specific tests.

Author: Actually, you are right

### E2E Test

> Verdaccio includes an User Interface and must be tested. We use End-to-End testing to run some smock test against the web API using the UI Theme 

Reviewer: ```suggestion
Verdaccio includes a Web User Interface that must be tested. We use End-to-End testing to run some smock tests against the web API using the UI Theme 
```

include by default.

```bash
    yarn lint && yarn test:all
```

The test does not have aim to test the integrity of the page, mostly, ensure the basic functionality still works. If you add or modify 
a UI feature, the tests must be updated.

> The tests rely on CSS classes naming convention, so, it is required some sort of coordination with the **verdaccio/ui** project.

We uses `puppeteer`, you can find more information about how to use it in their website. 

## Before commit

> We recommend run your tests and linters before commit.   

Reviewer: Just a thought - since we have husky already in the project we can go ahead and run tests before a push? it's less obtrusive than running them on each commit.

Author: It is easy to skip husky with `--no-verify` argument. For such reason I recommend it.

```bash
   yarn lint && yarn test:all
```

You can find more in our [guide about run and debugging test](https://github.com/verdaccio/verdaccio/wiki/Running-and-Debugging-tests#running-the-test).

## Continuous Integration

> Verdaccio uses [CircleCI](https://circleci.com/gh/verdaccio) as main Continuous Integration tool. We run the test against the most common 

Reviewer: ```suggestion
Verdaccio uses [CircleCI](https://circleci.com/gh/verdaccio) as its primary Continuous Integration tool. We run the tests against the most common Node.js versions available. Among them is LTS and the latest release. Before the PR is being merged, all checks must be green.
```

> Node.js versions available, LTS and the latest release. Before the PR is being merged, all check must be green.   

Reviewer: ```suggestion
```

Author: Why we remove this? I was not 100% happy when I wrote it, but now :) I'm curious.

Reviewer: I merged it with the paragraph and removed it from being it's own single line. That line of text is still there :-)

Author: Ah ok 👍 I saw it.

> ## How To Test Verdaccio

Reviewer 2: How to test Verdaccio

> Welcome to the testing folder at Verdaccio. This document aims to help you to go throug how Verdaccio show be tested.

Reviewer 2: throug => through
show => can (or do you mean should?)

Author: can :-) is perfect

> First of all we must explain the stack, we use `jest` and other tools as `supertest` to be able test the API and `puppetter` for End to End testing.

Reviewer 2: as => such as
to be able test => to be able to test
puppetter => puppeteer
End to End => End-to-End

> We go along with the following rules in order to be consisten with all test and will make your code review smooth and fast:

Reviewer 2: consisten => consistent
all test => all tests
and will make => which will make

> * We **type** all our test. eg `const foo: number = 3;`

Reviewer 2: test. => tests,

> * **Each test should be small as possible**: You should try the `test()` block try to test only one thing and no dependencies with other test. If test requires different steps, groupd them woth `describe()` block.

Reviewer 2: small as => as small as
block try => block
no dependencies with other test => do not depend on other tests
If test => If the test
groupd => group
woth => with a

> * All `test()` **headers title** should init with `test('should test ...)`: For consistency with reporting tools, this makes more easy to match the test with the feature need to be tested.

Reviewer 2: title => titles
init => begin
this makes more easy => this makes it easier
feature need => feature needed

* **Any mock data** should be located in the `partials` folder in each section.

> * Use `yaml` for **configuration files examples** instead use JSON.

Reviewer 2: instead use => instead of

> * If you use a **file based mock storage**, should be located in the `store` folder in each section.

Reviewer 2: should be => it should be

> * All test **MUST NOT** rely in external sources and must be able to run them **off line**.

Reviewer 2: test => tests
rely in => rely on
off line => offline

> * Test must run in the following Operative System: Unix (Mac, Linux) and Windows (7 -> latest).

Reviewer 2: Test => Tests
in => on
Operative System => Operating Systems / OS

> * If you are creating mock data file based and need a clean state, use `rimraf` to remove folders.

Reviewer 2: based and need a clean => which use the state and need a clean state

> ## Testing Sections

Reviewer 2: Testing => Test?
Sections => sections

> Verdaccio testing is being separated in 3 sections, each of them have different setup and scope, the most important is **unit test**. All sections has custom **jest configuration files**.

Reviewer 2: each of them have => each of them has
is unit => is the unit
All sections has => All sections have

> If you are adding new test, comply with the following:

Reviewer 2: adding new test => a new tests / new tests

* If you add a new API endpoint, the  unit test and functional test are mandatory.
* If you add a utility, unit test is mandatory.

> * If you are adding a new web API endpint, the unit test, functional test and if such endpoint has new changes in the UI, e2 is also mandatory.

Reviewer 2: e2 => E2E test

* If you add or refactor a core class, unit test is mandatory.

> * If you fix a bug, you **must** add a new `test()` block proving that the patch fix te bug.

Reviewer 2: proving that the patch fix te bug => to prove that the patch fixes the bug

> ### Unit Test

Reviewer 2: Test => test

> Unit test aims to test the CLI API and the Web API. The configuration file is located `jest.config.js`.

Reviewer 2: is located => is located at

> > Unit testing do not need pre-compile code, jest will catch any change done to the `{root}/src` files.

Reviewer 2: do not need pre-compile code => does not require pre-compiled code

#### Testing an endpoint

> We have prepared a template `test/unit/api/api.__test.template.spec.js` that you can follow to create your own unit test. Only the test are prefixed with `.spec.js` will be catch by jest.

Reviewer 2: prepared a template => prepared a template at
Only the test => Only the tests
prefixed with => appendend with
will be catch by => which will be found and used by

> Feel free to suggest improvements to the template, there is still a lot of room for improvement. 

We recommend the following approach when you create a unit test:

> * For new utilities, we recommend create a new spec.

Reviewer 2: create a new => creating a new / to create a new

* For existing utitlities, if the method is already being tested, just add a new `test()` block.

> * Notice that all api spec files are prefixed with `api.[feature].spec.js`, we recommend follow the same approach. eg: `api.[deprecate].spec.js`.

Reviewer 2: api => API
prefixed => appended
follow => following / to follow

> * Don't mix utilities with API test.

Reviewer 2: test => tests

> ### Functional Test

Reviewer 2: Test => tests

> The functional test aims to run only **cli endpoint** and **web point** real request to an existing and compiled verdaccio server running.

Reviewer 2: test aims to run only => tests aim to run only on the
real request => using real requests
verdaccio server running => running erdaccio server

> > Be aware if you change something in the `{root}/src` source code, you must run `yarn code:build` before be able see your changes, functional test uses transpiled code. 

Reviewer 2: before be able see your changes, functional test uses transpiled code => before to be able to see your changes because functional tests use the transpiled code.

> All tests must be included in the `test/functional/index.spec.js` file, wich bootstrap the whole process.

Reviewer 2: bootstrap => bootstraps

> The jest configuration file is defined in `test/jest.config.functional.js`. The configuration will create a custom environment launching 3 verdaccio servers with different configurations.

Reviewer 2: verdaccio => Verdaccio

> The servers are linked as follows. 

Reviewer 2: follows. => follows:

* Server 1 
 * -> Server 2  
 * -> Server 3
* Server 2 
 * -> Server 1
* Server 3 
  * -> Server 2
  * -> Server 1

Server 1 runs on port `55551`, Server 2 on port `55552` and Server 3 on port `55553`.

> > If you have the need to increase the number of servers running, it is possible, but please discuss with the team before go in that path.

Reviewer 2: before gon in that path => before you go that path

### E2E Test

> ## Before Commit

Reviewer 2: Commit => commit

> // set of steps we recommend to before commit

Reviewer 2: to before => to do before a

> ## Debugging Test

Reviewer 2: Debugging Test => Debug tests

> // debug processes is hard, we try to help you to find answers

Reviewer 2: debug => debugging

## Continous Integration

// we explain what we test in CI

> ## Run Test in Docker

Reviewer 2: Test => tests
